<div align="right" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/ExploreAI_logos/Logo blue_dark.png"  style="width:25px" align="right";/>
</div>

# Integrated Project: Maji Ndogo Part 2
© ExploreAI Academy

## Maji Ndogo: From analysis to action
# Clustering data to unveil Maji Ndogo's water crisis

> ⚠️ This notebook will not run on Google Colab because it cannot connect to a local database. Please make sure that this notebook is running on the same local machine as your MySQL Workbench installation and MySQL `mdd_water_services` database.




### Cleaning our data
Ok, bring up the employee table. It has info on all of our workers, but note that the email addresses have not been added. We will have to send them reports and figures, so let's update it. Luckily the emails for our department are easy: first_name.last_name@ndogowater.gov.

#### We can determine the email address for each employee by:
- selecting the employee_name column
- replacing the space with a full stop
- make it lowercase
- and stitch it all together

## Connecting to our MySQL database

Using our `Access_to_Basic_Services` table created in MySQL Workbench, we want to answer some questions on the range of our dataset. We can apply the same queries in MySQL Workbench and in this notebook if we connect to our MySQL server. Since we have a MySQL database, we can connect to it using mysql and pymysql.

In [ ]:
# To display all row in ouputs
%config SqlMagic.displaylimit = None

In [1]:
# Load and activate the SQL extension to allow us to execute SQL in a Jupyter notebook. 
# If you get an error here, make sure that mysql and pymysql are installed correctly. 

%load_ext sql

In [2]:
# Establish a connection to the local database using the '%sql' magic command.
# Replace 'password' with our connection password and `db_name` with our database name. 
# If you get an error here, please make sure the database name or password is correct.

%sql mysql+pymysql://Explore:712370SS@localhost:3306/mdd_water_services

Connecting to 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

In [3]:
%%sql
 -- Replace the space with a full stop AND Make it all lower case  PAGE (6)
SELECT
           LOWER(REPLACE(employee_name,' ','.'))               
FROM 
    employee

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

56 rows affected.

+---------------------------------------+
| LOWER(REPLACE(employee_name,' ','.')) |
+---------------------------------------+
|              amara.jengo              |
|              bello.azibo              |
|              bakari.iniko             |
|             malachi.mavuso            |
|              cheche.buhle             |
|             zuriel.matembo            |
|              deka.osumare             |
|             lalitha.kaburi            |
|              enitan.zuri              |
|               farai.nia               |
+---------------------------------------+
Truncated to displaylimit of 10.

In [4]:
%%sql
-- Replace the space with a full stop AND Make it all lower case,
-- and use CONCAT() '@ndogowater.gov') AS new_email −− add it all together   PAGE (6)

SELECT
     CONCAT(
           LOWER(REPLACE(employee_name,' ','.')), '@ndogowater.gov') AS new_email                    
FROM 
    employee

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

56 rows affected.

+-------------------------------+
|           new_email           |
+-------------------------------+
|   amara.jengo@ndogowater.gov  |
|   bello.azibo@ndogowater.gov  |
|  bakari.iniko@ndogowater.gov  |
| malachi.mavuso@ndogowater.gov |
|  cheche.buhle@ndogowater.gov  |
| zuriel.matembo@ndogowater.gov |
|  deka.osumare@ndogowater.gov  |
| lalitha.kaburi@ndogowater.gov |
|   enitan.zuri@ndogowater.gov  |
|    farai.nia@ndogowater.gov   |
+-------------------------------+
Truncated to displaylimit of 10.

In [5]:
%%sql

-- Use the employee table to count how many of our employees live in each town. 
-- Think carefully about what function we should use and how we should aggregate the data.   PAGE (9)

SELECT DISTINCT
       province_name, 
       town_name,
       COUNT(town_name)  AS num_employees 
 
 FROM
    mdd_water_services.employee 
GROUP BY 
      1,2
ORDER BY 
      1,3 DESC


Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

15 rows affected.

+---------------+-----------+---------------+
| province_name | town_name | num_employees |
+---------------+-----------+---------------+
|     Akatsi    |   Rural   |       5       |
|     Akatsi    |   Lusaka  |       4       |
|     Akatsi    |   Harare  |       3       |
|     Akatsi    |  Kintampo |       1       |
|     Amanzi    |   Dahabu  |       6       |
|     Amanzi    |   Rural   |       1       |
|    Hawassa    |   Rural   |       7       |
|    Hawassa    |  Zanzibar |       4       |
|    Hawassa    |   Serowe  |       3       |
|    Hawassa    |  Yaounde  |       1       |
+---------------+-----------+---------------+
Truncated to displaylimit of 10.

In [6]:
%%sql
-- Let's first look at the number of records each employee collected. So find the correct table, 
-- figure out what function to use and how to group, order
-- and limit the results to only see the top 3 employee_ids 
-- with the highest number of locations visited  PAGE (10)

SELECT 
     assigned_employee_id,
     COUNT(visit_count) AS number_of_visits                                                     
FROM
    mdd_water_services.visits
GROUP BY 
        assigned_employee_id
ORDER BY
       number_of_visits DESC
  

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

29 rows affected.

+----------------------+------------------+
| assigned_employee_id | number_of_visits |
+----------------------+------------------+
|          1           |       3708       |
|          30          |       3676       |
|          34          |       3539       |
|          3           |       3420       |
|          10          |       3407       |
|          8           |       3351       |
|          5           |       3284       |
|          36          |       3249       |
|          48          |       2933       |
|          28          |       2762       |
+----------------------+------------------+
Truncated to displaylimit of 10.

In [7]:
%%sql

-- Create a query that counts the number of records per town  PAGE (11)

SELECT DISTINCT 
     town_name,
     COUNT(town_name) OVER (PARTITION BY town_name) AS records_per_town
FROM
    mdd_water_services.location 

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

25 rows affected.

+-----------+------------------+
| town_name | records_per_town |
+-----------+------------------+
|  Abidjan  |       400        |
|   Amara   |       710        |
|   Amina   |       1090       |
|   Asmara  |       930        |
|   Bahari  |       470        |
|   Bello   |       400        |
|   Cheche  |       370        |
|   Dahabu  |       930        |
|    Deka   |       350        |
|   Djenne  |       310        |
+-----------+------------------+
Truncated to displaylimit of 10.

In [8]:
%%sql

-- Now count the records per province.  PAGE (12)
SELECT DISTINCT 
     province_name,
     COUNT(province_name) OVER (PARTITION BY province_name) AS records_per_province
FROM
    mdd_water_services.location  

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

5 rows affected.

+---------------+----------------------+
| province_name | records_per_province |
+---------------+----------------------+
|     Akatsi    |         8940         |
|     Amanzi    |         6950         |
|    Hawassa    |         6030         |
|    Kilimani   |         9510         |
|     Sokoto    |         8220         |
+---------------+----------------------+

### Can you find a way to do the following:   PAGE (13)
-- 1. Create a result set showing:
• province_name
• town_name
• An aggregated count of records for each town (consider naming this records_per_town).
• Ensure your data is grouped by both province_name and town_name.
2. Order your results primarily by province_name. Within each province, further sort the towns by their record counts in descending order.

In [9]:
%%sql
SELECT DISTINCT 
         province_name,
         town_name,
         COUNT(town_name) AS records_per_town
FROM
    mdd_water_services.location 
GROUP BY 
        province_name,
        town_name
ORDER BY province_name 

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

31 rows affected.

+---------------+-----------+------------------+
| province_name | town_name | records_per_town |
+---------------+-----------+------------------+
|     Akatsi    |   Harare  |       800        |
|     Akatsi    |  Kintampo |       780        |
|     Akatsi    |   Lusaka  |       1070       |
|     Akatsi    |   Rural   |       6290       |
|     Amanzi    |  Abidjan  |       400        |
|     Amanzi    |   Amina   |       670        |
|     Amanzi    |   Asmara  |       930        |
|     Amanzi    |   Bello   |       400        |
|     Amanzi    |   Dahabu  |       930        |
|     Amanzi    |   Pwani   |       520        |
+---------------+-----------+------------------+
Truncated to displaylimit of 10.

In [10]:
%%sql

-- Finally, look at the number of records for each location type   PAGE (13)

SELECT DISTINCT 
     location_type,
     COUNT(location_type) OVER (PARTITION BY location_type) AS num_sources
FROM
    mdd_water_services.location  

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

2 rows affected.

+---------------+-------------+
| location_type | num_sources |
+---------------+-------------+
|     Rural     |    23740    |
|     Urban     |    15910    |
+---------------+-------------+

In [11]:
%%sql

-- We can see that there are more rural sources than urban,
-- but it's really hard to understand those numbers. Percentages are more relatable.
-- If we use SQL as a very overpowered calculator:
SELECT
      23740 / (15910 + 23740) * 100 AS pct_rural_source
-- We can see that 60% of all water sources in the data set are in rural communities.

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

1 rows affected.

+------------------+
| pct_rural_source |
+------------------+
|     59.8739      |
+------------------+

So again, what are some of the insights we gained from the location table?
1. Our entire country was properly canvassed, and our dataset represents the situation on the ground.
2. 60% of our water sources are in rural communities across Maji Ndogo. We need to keep this in mind when we make decisions

In [12]:
%%sql

SELECT
      *
FROM
    mdd_water_services.water_source 
LIMIT 50

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

50 rows affected.

+--------------+----------------------+-------------------------+
|  source_id   | type_of_water_source | number_of_people_served |
+--------------+----------------------+-------------------------+
| AkHa00000224 |     tap_in_home      |           956           |
| AkHa00001224 |  tap_in_home_broken  |           930           |
| AkHa00002224 |  tap_in_home_broken  |           486           |
| AkHa00003224 |         well         |           364           |
| AkHa00004224 |  tap_in_home_broken  |           942           |
| AkHa00005224 |     tap_in_home      |           736           |
| AkHa00006224 |     tap_in_home      |           882           |
| AkHa00007224 |     tap_in_home      |           554           |
| AkHa00008224 |         well         |           398           |
| AkHa00009224 |         well         |           346           |
+--------------+----------------------+-------------------------+
Truncated to displaylimit of 10.

### The way I look at this table; 
we have access to different water source types and the number of people using each source.
These are the questions that I am curious about.
1. How many people did we survey in total?
2. How many wells, taps and rivers are there?
3. How many people share particular types of water sources on average?
4. How many people are getting water from each type of source? 

In [13]:
%%sql

-- 1. How many people did we survey in total?
SELECT
SUM(number_of_people_served) AS SURVEY_IN_TOTAL   -- > EQUAL = '27628140'
FROM
    mdd_water_services.water_source 

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

1 rows affected.

+-----------------+
| SURVEY_IN_TOTAL |
+-----------------+
|     27628140    |
+-----------------+

In [14]:
%%sql
-- 2. How many wells, taps and rivers are there? PAGE(15)

SELECT DISTINCT
     type_of_water_source,
     COUNT(type_of_water_source) OVER (PARTITION BY type_of_water_source) AS number_of_sources
FROM
    mdd_water_services.water_source 

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

5 rows affected.

+----------------------+-------------------+
| type_of_water_source | number_of_sources |
+----------------------+-------------------+
|        river         |        3379       |
|      shared_tap      |        5767       |
|     tap_in_home      |        7265       |
|  tap_in_home_broken  |        5856       |
|         well         |       17383       |
+----------------------+-------------------+

In [15]:
%%sql
-- 3. How many people share particular types of water sources on average?  PAGE(16)
SELECT DISTINCT
     type_of_water_source,
     ROUND(AVG(number_of_people_served) OVER (PARTITION BY type_of_water_source)) AS number_of_sources
FROM
    mdd_water_services.water_source  

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

5 rows affected.

+----------------------+-------------------+
| type_of_water_source | number_of_sources |
+----------------------+-------------------+
|        river         |        699        |
|      shared_tap      |        2071       |
|     tap_in_home      |        644        |
|  tap_in_home_broken  |        649        |
|         well         |        279        |
+----------------------+-------------------+

In [16]:
%%sql
-- Now let’s calculate the total number of people served by each type of water source in total, to make it easier to interpret, order them so the most people served by a source is at the top.   PAGE(19)

SELECT DISTINCT
     type_of_water_source,
     ROUND(SUM(number_of_people_served) OVER (PARTITION BY type_of_water_source)) AS population_served
FROM
    mdd_water_services.water_source  

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

5 rows affected.

+----------------------+-------------------+
| type_of_water_source | population_served |
+----------------------+-------------------+
|        river         |      2362544      |
|      shared_tap      |      11945272     |
|     tap_in_home      |      4678880      |
|  tap_in_home_broken  |      3799720      |
|         well         |      4841724      |
+----------------------+-------------------+

In [17]:
%%sql

-- Next, calculate the percentages using the total we just got.  
-- Let's round that off to 0 decimals, and order the results.   PAGE(21)

SELECT DISTINCT
     type_of_water_source,
     FORMAT((SUM(number_of_people_served)/27628140)*100,0) AS percentage_people_per_source
FROM
    mdd_water_services.water_source
GROUP BY
       type_of_water_source
ORDER BY
        percentage_people_per_source DESC
        

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

5 rows affected.

+----------------------+------------------------------+
| type_of_water_source | percentage_people_per_source |
+----------------------+------------------------------+
|        river         |              9               |
|      shared_tap      |              43              |
|         well         |              18              |
|     tap_in_home      |              17              |
|  tap_in_home_broken  |              14              |
+----------------------+------------------------------+

In [18]:
%%sql

-- So use a window function on the total people served column, converting it into a rank.    PAGE(23)
-- But think about this: If someone has a tap in their home, 
-- they already have the best source available. Since we can’t do anything more to improve
-- this, we should remove tap_in_home from the ranking before we continue.

SELECT DISTINCT
     type_of_water_source,
     SUM(number_of_people_served) AS people_served,
     RANK() OVER (ORDER BY SUM(number_of_people_served) DESC ) AS rank_by_population
FROM
    md_water_services.water_source
WHERE
    type_of_water_source != 'tap_in_home'
GROUP BY 
       type_of_water_source 
LIMIT 50

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

4 rows affected.

+----------------------+---------------+--------------------+
| type_of_water_source | people_served | rank_by_population |
+----------------------+---------------+--------------------+
|      shared_tap      |    11945272   |         1          |
|         well         |    4841724    |         2          |
|  tap_in_home_broken  |    3799720    |         3          |
|        river         |    2362544    |         4          |
+----------------------+---------------+--------------------+

In [19]:
%%sql

-- So create a query to do this, and keep these requirements in mind:   PAGE(24)
-- 1. The sources within each type should be assigned a rank.
-- 2. Limit the results to only improvable sources.
-- 3. Think about how to partition, filter and order the results set.
-- 4. Order the results to see the top of the list.
SELECT DISTINCT
     source_id,
     type_of_water_source,
     number_of_people_served,
     DENSE_RANK() OVER (ORDER BY number_of_people_served DESC ) AS priority_rank
FROM
    md_water_services.water_source
WHERE
    type_of_water_source != 'tap_in_home'  
LIMIT 50    


Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

50 rows affected.

+--------------+----------------------+-------------------------+---------------+
|  source_id   | type_of_water_source | number_of_people_served | priority_rank |
+--------------+----------------------+-------------------------+---------------+
| HaRu19509224 |      shared_tap      |           3998          |       1       |
| AkRu05603224 |      shared_tap      |           3998          |       1       |
| AkRu04862224 |      shared_tap      |           3996          |       2       |
| KiHa22867224 |      shared_tap      |           3996          |       2       |
| AmAs10911224 |      shared_tap      |           3996          |       2       |
| HaRu19839224 |      shared_tap      |           3994          |       3       |
| KiZu31330224 |      shared_tap      |           3994          |       3       |
| KiRu28630224 |      shared_tap      |           3992          |       4       |
| KiZu31415224 |      shared_tap      |           3992          |       4       |
| KiRu26218224 |      shared_tap      |           3990          |       5       |
+--------------+----------------------+-------------------------+---------------+
Truncated to displaylimit of 10.

### Ok, these are some of the things I think are worth looking at:
1. How long did the survey take?
2. What is the average total queue time for water?
3. What is the average queue time on different days?
4. How can we communicate this information efficiently?

In [20]:
%%sql

-- Question 1:   PAGE(27)
-- To calculate how long the survey took, 
-- we need to get the first and last dates (which functions can find the largest/smallest value), and subtract
-- them. Remember with DateTime data, 
-- we can't just subtract the values. We have to use a function to get the difference in days.
SELECT
    TIMESTAMPDIFF(DAY,MIN(time_of_record),MAX(time_of_record)) AS DURATION     
FROM
    mdd_water_services.visits;  

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

1 rows affected.

+----------+
| DURATION |
+----------+
|   924    |
+----------+

In [21]:
%%sql

-- Question 2:   PAGE(28)
-- Let's see how long people have to queue on average in Maji Ndogo. 
-- Keep in mind that many sources like taps_in_home have no queues. These
-- are just recorded as 0 in the time_in_queue column, 
-- so when we calculate averages, we need to exclude those rows. Try using NULLIF() do to this.
SELECT
      AVG(NULLIF(time_in_queue,0)) AS AVG_time_in_queue
FROM
    mdd_water_services.visits 
-- You should get a queue time of about 123 min. So on average, 
-- people take two hours to fetch water if they don't have a tap in their homes.

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

1 rows affected.

+-------------------+
| AVG_time_in_queue |
+-------------------+
|      123.2574     |
+-------------------+

In [22]:
%%sql

-- So let's look at the queue times aggregated across the different days of the week.     PAGE(29)

SELECT DISTINCT 
     DAYNAME(time_of_record) AS day_of_week, 
     ROUND(AVG(time_in_queue)) AS avg_queue_time
FROM 
   mdd_water_services.visits
GROUP BY day_of_week
ORDER BY day_of_week   

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

7 rows affected.

+-------------+----------------+
| day_of_week | avg_queue_time |
+-------------+----------------+
|    Friday   |       53       |
|    Monday   |       60       |
|   Saturday  |      246       |
|    Sunday   |       82       |
|   Thursday  |       46       |
|   Tuesday   |       47       |
|  Wednesday  |       43       |
+-------------+----------------+

In [23]:
%%sql

-- Question 4:    PAGE(31)
-- We can also look at what time during the day people collect water. 
-- Try to order the results in a meaningful way.

SELECT 
    TIME_FORMAT(TIME(time_of_record),'%H:00') AS hour_of_day, 
    ROUND(AVG(time_in_queue)) AS avg_queue_time
FROM 
     mdd_water_services.visits
GROUP BY 
     hour_of_day
ORDER BY 
      avg_queue_time
      
-- Can you see that mornings and evenings are the busiest? 
-- It looks like people collect water before and after work. 
-- Wouldn't it be nice to break down the queue times for each hour of each day? In a spreadsheet,
-- we can just create a pivot table.


Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

14 rows affected.

+-------------+----------------+
| hour_of_day | avg_queue_time |
+-------------+----------------+
|    11:00    |       46       |
|    12:00    |       47       |
|    13:00    |       47       |
|    14:00    |       47       |
|    16:00    |       47       |
|    10:00    |       48       |
|    15:00    |       48       |
|    09:00    |       49       |
|    18:00    |      147       |
|    07:00    |      149       |
+-------------+----------------+
Truncated to displaylimit of 10.

In [24]:
%%sql
# PAGE(33)

SELECT
      TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
      DAYNAME(time_of_record) AS DAY_NAME,
CASE
	WHEN DAYNAME(time_of_record) = 'Sunday' 
            THEN time_in_queue
		WHEN DAYNAME(time_of_record) = 'Monday' 
            THEN time_in_queue
		WHEN DAYNAME(time_of_record) = 'Saturday'
            THEN time_in_queue
		WHEN DAYNAME(time_of_record) = 'Tuesday'
            THEN time_in_queue
		WHEN DAYNAME(time_of_record) = 'Wednesday'
            THEN time_in_queue
		WHEN DAYNAME(time_of_record) = 'Thursday'
            THEN time_in_queue
		WHEN DAYNAME(time_of_record) = 'Friday'
            THEN time_in_queue
    ELSE NULL
END AS Time_of_waiting
FROM
    mdd_water_services.visits
WHERE
    time_in_queue != 0 
-- this exludes other sources with 0 queue times.
LIMIT 50

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

50 rows affected.

+-------------+----------+-----------------+
| hour_of_day | DAY_NAME | Time_of_waiting |
+-------------+----------+-----------------+
|    09:00    |  Friday  |        15       |
|    09:00    |  Friday  |        62       |
|    10:00    |  Friday  |        28       |
|    10:00    |  Friday  |        9        |
|    11:00    |  Friday  |        17       |
|    11:00    |  Friday  |       240       |
|    12:00    |  Friday  |        20       |
|    12:00    |  Friday  |       171       |
|    13:00    |  Friday  |        28       |
|    13:00    |  Friday  |        16       |
+-------------+----------+-----------------+
Truncated to displaylimit of 10.

In [25]:
%%sql

 -- creating a pivot table pf time waiting in each day   PAGE(35)
SELECT
TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
-- Sunday
ROUND(AVG(
	CASE
		WHEN DAYNAME(time_of_record) = 'Sunday'
              THEN time_in_queue
       ELSE NULL
    END),0) AS Sunday,
-- Monday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Monday' 
             THEN time_in_queue
       ELSE NULL
	END),0) AS Monday,
-- Tuesday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Tuesday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Tuesday,
-- Wednesday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Wednesday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Wednesday,
-- Thursday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Thursday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Thursday,
-- Friday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Friday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Friday,
-- Saturday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Saturday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Saturday
FROM
   mdd_water_services.visits
WHERE
   time_in_queue != 0 -- this excludes other sources with 0 queue times
GROUP BY
     hour_of_day
ORDER BY
     hour_of_day;


Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

14 rows affected.

+-------------+--------+--------+---------+-----------+----------+--------+----------+
| hour_of_day | Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
|    06:00    |   79   |  190   |   134   |    112    |   134    |  153   |   247    |
|    07:00    |   82   |  186   |   128   |    111    |   139    |  156   |   247    |
|    08:00    |   86   |  183   |   130   |    119    |   129    |  153   |   247    |
|    09:00    |   84   |  127   |   105   |     94    |    99    |  107   |   252    |
|    10:00    |   83   |  119   |    99   |     89    |    95    |  112   |   259    |
|    11:00    |   78   |  115   |   102   |     86    |    99    |  104   |   236    |
|    12:00    |   78   |  115   |    97   |     88    |    96    |  109   |   239    |
|    13:00    |   81   |  122   |    97   |     98    |   101    |  115   |   242    |
|    14:00    |   83   |  127   |   104   |     92    |    96    |  110   |   244    |
|    15:00    |   83   |  126   |   104   |     88    |    92    |  110   |   248    |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
Truncated to displaylimit of 10.

# Questions of Integrated Project: Maji Ndogo Part 2

In [26]:
%%sql

-- Q1 

-- Which SQL query will produce the date format "DD Month YYYY" 
-- from the time_of_record column in the visits table, 
-- as a single column? Note: Monthname() acts in a similar way to DAYNAME().

-- ------------------------------------> ANSWER: This below query
SELECT 
    CONCAT(day(time_of_record), " ", monthname(time_of_record), " ", year(time_of_record)) AS Standard_date
FROM 
   mdd_water_services.visits
LIMIT 50

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

50 rows affected.

+----------------+
| Standard_date  |
+----------------+
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
+----------------+
Truncated to displaylimit of 10.

### Q2

You are working with an SQL query designed 

to calculate the Annual Rate of Change (ARC) for basic rural water services:

SELECT
    name,
    wat_bas_r - LAG(wat_bas_r) OVER (PARTITION BY (a) ORDER BY (b)) 
FROM 
    global_water_access
ORDER BY
     name;

To accomplish this task, what should you use for placeholders (a) and (b)?

-- ------------------------------------> ANSWER: Use name for grouping and year for sorting.

In [27]:
%%sql

-- Q2

SELECT
    name,
    wat_bas_r - LAG(wat_bas_r) OVER (PARTITION BY name  ORDER BY year) AS ARC
FROM 
   mdd_water_services.global_water_access
ORDER BY
     name;

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

464 rows affected.

+----------------+--------------------+
|      name      |        ARC         |
+----------------+--------------------+
|  Afghanistan   |        None        |
|  Afghanistan   | 13.339405059814453 |
|    Albania     |        None        |
|    Albania     | 3.4641036987304688 |
|    Algeria     |        None        |
|    Algeria     | 1.6847991943359375 |
| American Samoa |        None        |
| American Samoa |        None        |
|    Andorra     |        None        |
|    Andorra     |        0.0         |
+----------------+--------------------+
Truncated to displaylimit of 10.

### Q3
What are the names of the two worst-performing employees who visited the fewest sites,

and how many sites did the worst-performing employee visit? 

Modify your queries from the “Honouring the workers” section

###  ------------------------------------> ANSWER: Lesedi Kofi, Kunto Asha (15)
#### Below is Justifying my Answer

In [28]:
%%sql

-- Q3
SELECT 
     assigned_employee_id,
     COUNT(visit_count) AS number_of_visits                                                 
FROM
    mdd_water_services.visits
GROUP BY 
        assigned_employee_id
ORDER BY
       number_of_visits ASC
LIMIT 2;

 


Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

2 rows affected.

+----------------------+------------------+
| assigned_employee_id | number_of_visits |
+----------------------+------------------+
|          20          |        15        |
|          22          |       143        |
+----------------------+------------------+

In [29]:
%%sql

-- Q3

-- next do it 
select
     assigned_employee_id,
     employee_name
from
    mdd_water_services.employee 
WHERE
    assigned_employee_id = 20
OR
   assigned_employee_id =  22

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

2 rows affected.

+----------------------+---------------+
| assigned_employee_id | employee_name |
+----------------------+---------------+
|          20          |   Kunto Asha  |
|          22          |  Lesedi Kofi  |
+----------------------+---------------+

In [30]:
%%sql

-- Q4

-- What does the following query do?
SELECT 
    location_id,
    time_in_queue,
    visit_count,
    AVG(time_in_queue) OVER (PARTITION BY location_id ORDER BY visit_count) AS total_avg_queue_time
FROM 
    mdd_water_services.visits
WHERE 
     visit_count > 1 -- Only shared taps were visited > 1
ORDER BY 
    location_id, 
    time_of_record
LIMIT 50
    
-- ------------> ANSWER : It computes an average queue time for shared taps visited more than once,
              -- which is updated each time a source is visited.

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

50 rows affected.

+-------------+---------------+-------------+----------------------+
| location_id | time_in_queue | visit_count | total_avg_queue_time |
+-------------+---------------+-------------+----------------------+
|  AkHa00036  |      113      |      2      |       113.0000       |
|  AkHa00036  |      229      |      3      |       171.0000       |
|  AkHa00036  |      400      |      4      |       247.3333       |
|  AkHa00036  |      120      |      5      |       215.5000       |
|  AkHa00036  |      185      |      6      |       209.4000       |
|  AkHa00036  |      158      |      7      |       200.8333       |
|  AkHa00036  |      226      |      8      |       204.4286       |
|  AkHa00090  |       50      |      2      |       50.0000        |
|  AkHa00090  |       63      |      3      |       56.5000        |
|  AkHa00090  |       12      |      4      |       41.6667        |
+-------------+---------------+-------------+----------------------+
Truncated to displaylimit of 10.

In [31]:
%%sql

-- Q5
-- One of our employees, Farai Nia,
-- lives at 33 Angelique Kidjo Avenue. What would be the result if we TRIM() her address?
-- TRIM('33 Angelique Kidjo Avenue  ')

SELECT 
 TRIM('33 Angelique Kidjo Avenue  ') AS RESULT

-- ------------------------------------> ANSWER: ‘33 Angelique Kidjo Avenue’


Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

1 rows affected.

+---------------------------+
|           RESULT          |
+---------------------------+
| 33 Angelique Kidjo Avenue |
+---------------------------+

In [32]:
%%sql

 -- Q6  
-- How many employees live in Dahabu? 
-- Rely on one of the queries we used in the project to answer this.

SELECT DISTINCT
      town_name,
      COUNT(town_name) OVER (PARTITION BY town_name) AS num_employees                          
    
FROM
    mdd_water_services.employee 
WHERE
     town_name = 'Dahabu'
-- ------------------------------------> ANSWER: 6

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

1 rows affected.

+-----------+---------------+
| town_name | num_employees |
+-----------+---------------+
|   Dahabu  |       6       |
+-----------+---------------+

In [33]:
%%sql

-- Q7
-- How many employees live in Harare, Kilimani?
-- Modify one of your queries from the project to answer this question.

SELECT DISTINCT
       province_name, 
       town_name,
       COUNT(town_name)  AS num_employees 
 
 FROM
    mdd_water_services.employee 
WHERE
     province_name = 'Kilimani'
AND  town_name = 'Harare'
GROUP BY 
      1,2
ORDER BY 
      1,3 DESC
-- ------------------------------------> ANSWER: 2

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

1 rows affected.

+---------------+-----------+---------------+
| province_name | town_name | num_employees |
+---------------+-----------+---------------+
|    Kilimani   |   Harare  |       2       |
+---------------+-----------+---------------+

In [34]:
%%sql

-- Q8
-- How many people share a well on average? Round your answer to 0 decimals.

SELECT DISTINCT
     type_of_water_source,
     ROUND(AVG(number_of_people_served) OVER (PARTITION BY type_of_water_source)) AS AVG_number_of_people_served
FROM
    mdd_water_services.water_source 
WHERE
    type_of_water_source = 'well'
-- ------------------------------------> ANSWER: 279


Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

1 rows affected.

+----------------------+-----------------------------+
| type_of_water_source | AVG_number_of_people_served |
+----------------------+-----------------------------+
|         well         |             279             |
+----------------------+-----------------------------+

### Q9

Consider the query we used to calculate the total number of people served:

SELECT
    SUM(number_of_people_served) AS population_served
FROM
   water_source
ORDER BY
    population_served

Which of the following lines of code will calculate the total number of people using some sort of tap?

### -- ------------------------------------> ANSWER: WHERE type_of_water_source LIKE "%tap%"

#### Below is Justifying my Answer

In [35]:
%%sql

-- Q9 

SELECT
     type_of_water_source,
     SUM(number_of_people_served) AS population_served
FROM
   md_water_services.water_source
WHERE 
    type_of_water_source LIKE "%tap%"
GROUP BY
    type_of_water_source
ORDER BY
    population_served DESC


Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

3 rows affected.

+----------------------+-------------------+
| type_of_water_source | population_served |
+----------------------+-------------------+
|      shared_tap      |      11945272     |
|     tap_in_home      |      4678880      |
|  tap_in_home_broken  |      3799720      |
+----------------------+-------------------+

### Q10

Use the pivot table we created to answer the following question. What are the average queue times for the following times?

Saturday from 12:00 to 13:00
Tuesday from 18:00 to 19:00
Sunday from 09:00 to 10:00



### -- ------------------------------------> ANSWER: Saturday: 239, Tuesday: 122, Sunday: 84

#### Below is Justifying my Answer

In [36]:
%%sql

-- Q10

SELECT
TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
-- Sunday
ROUND(AVG(
	CASE
		WHEN DAYNAME(time_of_record) = 'Sunday'
              THEN time_in_queue
       ELSE NULL
    END),0) AS Sunday,
-- Monday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Monday' 
             THEN time_in_queue
       ELSE NULL
	END),0) AS Monday,
-- Tuesday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Tuesday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Tuesday,
-- Wednesday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Wednesday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Wednesday,
-- Thursday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Thursday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Thursday,
-- Friday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Friday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Friday,
-- Saturday
ROUND(AVG(
	CASE
       WHEN DAYNAME(time_of_record) = 'Saturday'
             THEN time_in_queue
       ELSE NULL
	END),0) AS Saturday
FROM
   md_water_services.visits
WHERE
   time_in_queue != 0 -- this excludes other sources with 0 queue times
GROUP BY
     hour_of_day
ORDER BY
     hour_of_day;

Running query in 'mysql+pymysql://Explore:***@localhost:3306/mdd_water_services'

14 rows affected.

+-------------+--------+--------+---------+-----------+----------+--------+----------+
| hour_of_day | Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
|    06:00    |   79   |  190   |   134   |    112    |   134    |  153   |   247    |
|    07:00    |   82   |  186   |   128   |    111    |   139    |  156   |   247    |
|    08:00    |   86   |  183   |   130   |    119    |   129    |  153   |   247    |
|    09:00    |   84   |  127   |   105   |     94    |    99    |  107   |   252    |
|    10:00    |   83   |  119   |    99   |     89    |    95    |  112   |   259    |
|    11:00    |   78   |  115   |   102   |     86    |    99    |  104   |   236    |
|    12:00    |   78   |  115   |    97   |     88    |    96    |  109   |   239    |
|    13:00    |   81   |  122   |    97   |     98    |   101    |  115   |   242    |
|    14:00    |   83   |  127   |   104   |     92    |    96    |  110   |   244    |
|    15:00    |   83   |  126   |   104   |     88    |    92    |  110   |   248    |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
Truncated to displaylimit of 10.

#  
<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/ExploreAI_logos/EAI_Blue_Dark.png"  style="width:200px";/>
</div>